File: implementations/java/client/src/main/java/com/theshuai/specusclient/peer/PeerEgressRouteInstaller.java
package com.theshuai.specusclient.peer;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.theshuai.specusclient.peer.PeerEgressRoutePlanner.Kind;
import com.theshuai.specusclient.peer.PeerEgressRoutePlanner.Route;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * Installing routes without leaving any behind.
 *
 * <p>Two problems this class exists to solve. A partial install has to come back out, or a failure
 * halfway through a rule set leaves the machine routing some traffic into a tunnel that was never
 * finished being set up. And a route this feature owns has to be distinguishable from one the user
 * or another tool installed, across a clean exit, a crash and a restart, because withdrawing
 * someone else's route is worse than leaving our own behind.
 *
 * <p>The journal is what answers the second question. It records what was installed, on disk,
 * before the install is attempted. Written first on purpose: a journal entry for a route that
 * failed to install costs one harmless removal attempt at cleanup, while a route installed without
 * a journal entry is one nobody will ever take back.
 *
 * <p>Its on-disk shape is shared with the Go and .NET consumers, and pinned by
 * {@code protocol/test-vectors/peer-egress-routes-v1.json}. A user who switches implementations on
 * one machine has to have their routes adopted and withdrawn rather than left behind by a reader
 * that did not recognise the file.
 *
 * <p>Not safe for concurrent use; the caller that owns the rule configuration drives it.
 */
public final class PeerEgressRouteInstaller {

    static final int JOURNAL_VERSION = 1;

    private static final ObjectMapper MAPPER = new ObjectMapper();

    /**
     * The platform's routing table. Injected so the failure and rollback paths can be driven
     * without touching the real one.
     */
    public interface Commander {
        /**
         * Reports an existing route for this exact prefix that this feature does not own. The
         * description is what an operator reads, so it should say what is already there.
         */
        Conflict conflict(Route route);

        void install(Route route) throws IOException;

        void remove(Route route) throws IOException;
    }

    /** Whether a prefix is already taken, and by what. */
    public record Conflict(boolean present, String existing) {
        private static final Conflict NONE = new Conflict(false, "");

        public static Conflict none() {
            return NONE;
        }
    }

    /** A route refused because something else already owns the prefix. */
    public record RouteConflict(Route route, String existing) {
    }

    /**
     * What happened, in enough detail for an operator to act.
     *
     * @param conflicts routes refused because something else already owns the prefix; the rest of
     *                  the plan is applied regardless, since one contested prefix should not
     *                  disable every other rule
     * @param rolledBack set when an install failed for a reason other than a conflict and the
     *                   additions from this call were undone
     */
    public record ApplyResult(
            List<Route> added,
            List<Route> removed,
            List<RouteConflict> conflicts,
            boolean rolledBack,
            Exception error) {
    }

    private final Commander commander;
    private final Path journalPath;
    private final List<Route> installed = new ArrayList<>();

    public PeerEgressRouteInstaller(Commander commander, Path journalPath) {
        this.commander = commander;
        this.journalPath = journalPath;
    }

    /** The routes this feature currently owns, in the order they were installed. */
    public List<Route> installed() {
        return List.copyOf(installed);
    }

    /**
     * Reads the journal from a previous run.
     *
     * <p>A missing journal yields an empty set: there is nothing to be gained by refusing to start,
     * and the alternative to an empty set is guessing which of the machine's routes might have been
     * ours. A journal that exists but cannot be read is an error rather than an empty set, because
     * treating it as empty would mean the routes it describes are never taken back at all.
     */
    public void load() throws IOException {
        installed.clear();
        if (journalPath == null) {
            return;
        }
        if (!Files.exists(journalPath)) {
            return;
        }
        JsonNode journal = MAPPER.readTree(Files.readString(journalPath, StandardCharsets.UTF_8));
        if (!journal.isObject()) {
            throw new IOException("egress route journal is not an object");
        }
        int version = journal.path("version").asInt(-1);
        if (version != JOURNAL_VERSION) {
            // A journal from a version that wrote entries differently cannot be trusted to
            // describe what is actually installed, and acting on it would mean removing prefixes
            // by guess.
            throw new IOException("egress route journal version " + version + " is not " + JOURNAL_VERSION);
        }
        List<Route> adopted = new ArrayList<>();
        for (JsonNode node : journal.path("routes")) {
            String cidr = node.path("cidr").asText("");
            String kind = node.path("kind").asText("");
            if (cidr.isEmpty()) {
                throw new IOException("egress route journal carries a route with no prefix");
            }
            if (!"tun".equals(kind) && !"bypass".equals(kind)) {
                // Refused rather than defaulted: guessing would mean withdrawing a prefix the
                // previous run may have installed as the other kind.
                throw new IOException("unknown egress route kind " + kind);
            }
            adopted.add(new Route(cidr, Kind.fromWireName(kind), node.path("origin").asText("")));
        }
        PeerEgressRoutePlanner.sort(adopted);
        installed.addAll(adopted);
    }

    private void save() throws IOException {
        if (journalPath == null) {
            return;
        }
        if (installed.isEmpty()) {
            Files.deleteIfExists(journalPath);
            return;
        }
        PeerEgressRoutePlanner.sort(installed);
        // Written through a temporary file and renamed into place: a journal truncated by a crash
        // mid-write would describe fewer routes than are installed, and the difference is what gets
        // left behind.
        SecretFileWriter.writeSecret(journalPath, render(installed));
    }

    /**
     * Renders the journal.
     *
     * <p>Assembled by hand rather than through a pretty-printer so that the bytes are the ones the
     * shared vector pins, whatever a JSON library's default spacing happens to be this year. The
     * values still go through the encoder, because an origin carries a rule's match string and a
     * malformed configuration can put a quote in it.
     */
    static String render(List<Route> routes) throws IOException {
        StringBuilder text = new StringBuilder(128);
        text.append("{\n  \"version\": ").append(JOURNAL_VERSION).append(",\n  \"routes\": [");
        for (int index = 0; index < routes.size(); index++) {
            Route route = routes.get(index);
            text.append(index == 0 ? "\n" : ",\n");
            text.append("    {\n      \"cidr\": ").append(MAPPER.writeValueAsString(route.cidr()))
                    .append(",\n      \"kind\": ").append(MAPPER.writeValueAsString(route.kind().wireName()))
                    .append(",\n      \"origin\": ").append(MAPPER.writeValueAsString(route.origin()))
                    .append("\n    }");
        }
        text.append(routes.isEmpty() ? "]" : "\n  ]").append("\n}");
        return text.toString();
    }

    /**
     * Moves the routing table to the desired set.
     *
     * <p>Withdrawals happen first, so a prefix that changed kind loses its old entry before the new
     * one goes in. Then additions, each preceded by a conflict check.
     */
    public ApplyResult apply(List<Route> desired) {
        List<Route> added = new ArrayList<>();
        List<Route> removed = new ArrayList<>();
        List<RouteConflict> conflicts = new ArrayList<>();
        Exception error = null;

        PeerEgressRoutePlanner.Difference difference = PeerEgressRoutePlanner.diff(installed(), desired);
        for (Route route : difference.remove()) {
            try {
                commander.remove(route);
            } catch (Exception failure) {
                // A route we cannot remove is dropped from the journal anyway. Keeping it would
                // mean retrying forever against a table that no longer has it, which is the more
                // likely explanation than a table refusing us.
                error = failure;
            }
            forget(route);
            removed.add(route);
        }

        List<Route> appliedThisCall = new ArrayList<>();
        for (Route route : difference.add()) {
            Conflict conflict = commander.conflict(route);
            if (conflict.present()) {
                // Phase one does not preempt and does not compare metrics. Refusing one prefix and
                // telling the operator beats quietly winning an argument with their own routing.
                conflicts.add(new RouteConflict(route, conflict.existing()));
                continue;
            }
            // Journal before install, so a crash between the two leaves a removable record rather
            // than an unowned route.
            installed.add(route);
            try {
                save();
            } catch (Exception failure) {
                forget(route);
                rollback(appliedThisCall);
                return new ApplyResult(List.of(), List.copyOf(removed), List.copyOf(conflicts), true, failure);
            }
            try {
                commander.install(route);
            } catch (Exception failure) {
                forget(route);
                saveQuietly();
                rollback(appliedThisCall);
                return new ApplyResult(List.of(), List.copyOf(removed), List.copyOf(conflicts), true, failure);
            }
            appliedThisCall.add(route);
            added.add(route);
        }

        try {
            save();
        } catch (Exception failure) {
            if (error == null) {
                error = failure;
            }
        }
        return new ApplyResult(List.copyOf(added), List.copyOf(removed), List.copyOf(conflicts), false, error);
    }

    /**
     * Withdraws the additions from the current call only.
     *
     * <p>Only this call's, because routes from earlier calls are still wanted; a failure now is not
     * a reason to tear down a configuration that was working a minute ago.
     */
    private void rollback(List<Route> applied) {
        for (int index = applied.size() - 1; index >= 0; index--) {
            Route route = applied.get(index);
            try {
                commander.remove(route);
            } catch (Exception ignored) {
                // Already failing; the journal entry is what makes a retry possible.
            }
            forget(route);
        }
        saveQuietly();
    }

    /**
     * Removes every route this feature owns, for shutdown or for a restart that found a journal
     * from a previous run.
     */
    public List<Route> withdrawAll() {
        List<Route> owned = new ArrayList<>(installed);
        // Reverse order, so the bypass entries that keep the transport working outlive the routes
        // that point into the tunnel.
        PeerEgressRoutePlanner.sort(owned);
        for (int index = owned.size() - 1; index >= 0; index--) {
            try {
                commander.remove(owned.get(index));
            } catch (Exception ignored) {
                // Shutdown path: a route the table no longer has is the common reason to fail.
            }
        }
        installed.clear();
        saveQuietly();
        return owned;
    }

    /**
     * Reports whether a prefix is one of ours, which is what keeps a conflict check from treating
     * this feature's own route as somebody else's.
     */
    public boolean owns(String cidr) {
        for (Route route : installed) {
            if (route.cidr().equals(cidr)) {
                return true;
            }
        }
        return false;
    }

    private void forget(Route route) {
        installed.removeIf(existing -> existing.cidr().equals(route.cidr()));
    }

    private void saveQuietly() {
        try {
            save();
        } catch (Exception ignored) {
            // The caller is already on a failure path and has an error to report.
        }
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerEgressRouteInstaller.cs
using System.Text;
using System.Text.Json;

namespace Specus.Client.PeerMesh;

/// <summary>Whether a prefix is already taken, and by what.</summary>
internal readonly record struct PeerEgressRouteConflictCheck(bool Present, string Existing)
{
    public static PeerEgressRouteConflictCheck None { get; } = new(false, string.Empty);
}

/// <summary>A route refused because something else already owns the prefix.</summary>
internal readonly record struct PeerEgressRouteConflict(PeerEgressRoute Route, string Existing);

/// <summary>
/// What happened, in enough detail for an operator to act.
/// </summary>
/// <param name="Conflicts">
/// Routes refused because something else already owns the prefix. The rest of the plan is applied
/// regardless: one contested prefix should not disable every other rule.
/// </param>
/// <param name="RolledBack">
/// Set when an install failed for a reason other than a conflict and the additions from this call
/// were undone.
/// </param>
internal sealed record PeerEgressRouteApplyResult(
    IReadOnlyList<PeerEgressRoute> Added,
    IReadOnlyList<PeerEgressRoute> Removed,
    IReadOnlyList<PeerEgressRouteConflict> Conflicts,
    bool RolledBack,
    Exception? Error);

/// <summary>
/// The platform's routing table. Injected so the failure and rollback paths can be driven without
/// touching the real one.
/// </summary>
internal interface IPeerEgressRouteCommander
{
    /// <summary>
    /// Reports an existing route for this exact prefix that this feature does not own. The
    /// description is what an operator reads, so it should say what is already there.
    /// </summary>
    PeerEgressRouteConflictCheck Conflict(PeerEgressRoute route);

    void Install(PeerEgressRoute route);

    void Remove(PeerEgressRoute route);
}

/// <summary>
/// Installing routes without leaving any behind.
/// </summary>
/// <remarks>
/// Two problems this class exists to solve. A partial install has to come back out, or a failure
/// halfway through a rule set leaves the machine routing some traffic into a tunnel that was never
/// finished being set up. And a route this feature owns has to be distinguishable from one the user
/// or another tool installed, across a clean exit, a crash and a restart, because withdrawing
/// someone else's route is worse than leaving our own behind.
///
/// <para>The journal is what answers the second question. It records what was installed, on disk,
/// before the install is attempted. Written first on purpose: a journal entry for a route that
/// failed to install costs one harmless removal attempt at cleanup, while a route installed without
/// a journal entry is one nobody will ever take back.</para>
///
/// <para>Its on-disk shape is shared with the Go and Java consumers, and pinned by
/// <c>protocol/test-vectors/peer-egress-routes-v1.json</c>. A user who switches implementations on
/// one machine has to have their routes adopted and withdrawn rather than left behind by a reader
/// that did not recognise the file.</para>
///
/// <para>Not safe for concurrent use; the caller that owns the rule configuration drives it.</para>
/// </remarks>
internal sealed class PeerEgressRouteInstaller(IPeerEgressRouteCommander commander, string? journalPath)
{
    public const int JournalVersion = 1;

    private readonly List<PeerEgressRoute> _installed = [];

    /// <summary>The routes this feature currently owns.</summary>
    public IReadOnlyList<PeerEgressRoute> Installed => _installed;

    /// <summary>
    /// Reads the journal from a previous run.
    /// </summary>
    /// <remarks>
    /// A missing journal yields an empty set: there is nothing to be gained by refusing to start,
    /// and the alternative to an empty set is guessing which of the machine's routes might have
    /// been ours. A journal that exists but cannot be read throws rather than yielding an empty
    /// set, because treating it as empty would mean the routes it describes are never taken back.
    /// </remarks>
    public void Load()
    {
        _installed.Clear();
        if (string.IsNullOrEmpty(journalPath) || !File.Exists(journalPath))
        {
            return;
        }

        using var journal = JsonDocument.Parse(File.ReadAllText(journalPath));
        var root = journal.RootElement;
        if (root.ValueKind != JsonValueKind.Object)
        {
            throw new InvalidDataException("egress route journal is not an object");
        }
        var version = root.TryGetProperty("version", out var declared) && declared.TryGetInt32(out var value)
            ? value
            : -1;
        if (version != JournalVersion)
        {
            // A journal from a version that wrote entries differently cannot be trusted to describe
            // what is actually installed, and acting on it would mean removing prefixes by guess.
            throw new InvalidDataException(
                $"egress route journal version {version} is not {JournalVersion}");
        }

        var adopted = new List<PeerEgressRoute>();
        if (root.TryGetProperty("routes", out var routes) && routes.ValueKind == JsonValueKind.Array)
        {
            foreach (var node in routes.EnumerateArray())
            {
                var cidr = node.TryGetProperty("cidr", out var prefix) ? prefix.GetString() : null;
                var kind = node.TryGetProperty("kind", out var name) ? name.GetString() : null;
                if (string.IsNullOrEmpty(cidr))
                {
                    throw new InvalidDataException("egress route journal carries a route with no prefix");
                }
                if (kind is not ("tun" or "bypass"))
                {
                    // Refused rather than defaulted: guessing would mean withdrawing a prefix the
                    // previous run may have installed as the other kind.
                    throw new InvalidDataException($"unknown egress route kind {kind}");
                }
                var origin = node.TryGetProperty("origin", out var source) ? source.GetString() : null;
                adopted.Add(new PeerEgressRoute(
                    cidr, PeerEgressRoutePlanner.KindFromWireName(kind), origin ?? string.Empty));
            }
        }
        PeerEgressRoutePlanner.Sort(adopted);
        _installed.AddRange(adopted);
    }

    private void Save()
    {
        if (string.IsNullOrEmpty(journalPath))
        {
            return;
        }
        if (_installed.Count == 0)
        {
            File.Delete(journalPath);
            return;
        }
        PeerEgressRoutePlanner.Sort(_installed);
        // Written through a temporary file and renamed into place: a journal truncated by a crash
        // mid-write would describe fewer routes than are installed, and the difference is what gets
        // left behind.
        SecretFileWriter.WriteSecret(journalPath, Render(_installed));
    }

    /// <summary>
    /// Renders the journal.
    /// </summary>
    /// <remarks>
    /// Assembled by hand rather than through a serializer's indentation settings, so that the bytes
    /// are the ones the shared vector pins whatever a JSON library's default spacing happens to be
    /// this year. The values still go through the encoder, because an origin carries a rule's match
    /// string.
    /// </remarks>
    public static string Render(IReadOnlyList<PeerEgressRoute> routes)
    {
        var text = new StringBuilder(128);
        text.Append("{\n  \"version\": ").Append(JournalVersion).Append(",\n  \"routes\": [");
        for (var index = 0; index < routes.Count; index++)
        {
            var route = routes[index];
            text.Append(index == 0 ? "\n" : ",\n")
                .Append("    {\n      \"cidr\": ").Append(JsonSerializer.Serialize(route.Cidr))
                .Append(",\n      \"kind\": ")
                .Append(JsonSerializer.Serialize(PeerEgressRoutePlanner.WireName(route.Kind)))
                .Append(",\n      \"origin\": ").Append(JsonSerializer.Serialize(route.Origin))
                .Append("\n    }");
        }
        return text.Append(routes.Count == 0 ? "]" : "\n  ]").Append("\n}").ToString();
    }

    /// <summary>
    /// Moves the routing table to the desired set.
    /// </summary>
    /// <remarks>
    /// Withdrawals happen first, so a prefix that changed kind loses its old entry before the new
    /// one goes in. Then additions, each preceded by a conflict check.
    /// </remarks>
    public PeerEgressRouteApplyResult Apply(IReadOnlyList<PeerEgressRoute> desired)
    {
        var added = new List<PeerEgressRoute>();
        var removed = new List<PeerEgressRoute>();
        var conflicts = new List<PeerEgressRouteConflict>();
        Exception? error = null;

        var difference = PeerEgressRoutePlanner.Diff([.. _installed], desired);
        foreach (var route in difference.Remove)
        {
            try
            {
                commander.Remove(route);
            }
            catch (Exception failure)
            {
                // A route we cannot remove is dropped from the journal anyway. Keeping it would
                // mean retrying forever against a table that no longer has it, which is the more
                // likely explanation than a table refusing us.
                error = failure;
            }
            Forget(route);
            removed.Add(route);
        }

        var appliedThisCall = new List<PeerEgressRoute>();
        foreach (var route in difference.Add)
        {
            var conflict = commander.Conflict(route);
            if (conflict.Present)
            {
                // Phase one does not preempt and does not compare metrics. Refusing one prefix and
                // telling the operator beats quietly winning an argument with their own routing.
                conflicts.Add(new PeerEgressRouteConflict(route, conflict.Existing));
                continue;
            }
            // Journal before install, so a crash between the two leaves a removable record rather
            // than an unowned route.
            _installed.Add(route);
            try
            {
                Save();
            }
            catch (Exception failure)
            {
                Forget(route);
                Rollback(appliedThisCall);
                return new PeerEgressRouteApplyResult([], removed, conflicts, true, failure);
            }
            try
            {
                commander.Install(route);
            }
            catch (Exception failure)
            {
                Forget(route);
                SaveQuietly();
                Rollback(appliedThisCall);
                return new PeerEgressRouteApplyResult([], removed, conflicts, true, failure);
            }
            appliedThisCall.Add(route);
            added.Add(route);
        }

        try
        {
            Save();
        }
        catch (Exception failure)
        {
            error ??= failure;
        }
        return new PeerEgressRouteApplyResult(added, removed, conflicts, false, error);
    }

    /// <summary>
    /// Withdraws the additions from the current call only.
    /// </summary>
    /// <remarks>
    /// Only this call's, because routes from earlier calls are still wanted; a failure now is not a
    /// reason to tear down a configuration that was working a minute ago.
    /// </remarks>
    private void Rollback(List<PeerEgressRoute> applied)
    {
        for (var index = applied.Count - 1; index >= 0; index--)
        {
            try
            {
                commander.Remove(applied[index]);
            }
            catch (Exception)
            {
                // Already failing; the journal entry is what makes a retry possible.
            }
            Forget(applied[index]);
        }
        SaveQuietly();
    }

    /// <summary>
    /// Removes every route this feature owns, for shutdown or for a restart that found a journal
    /// from a previous run.
    /// </summary>
    public IReadOnlyList<PeerEgressRoute> WithdrawAll()
    {
        var owned = new List<PeerEgressRoute>(_installed);
        // Reverse order, so the bypass entries that keep the transport working outlive the routes
        // that point into the tunnel.
        PeerEgressRoutePlanner.Sort(owned);
        for (var index = owned.Count - 1; index >= 0; index--)
        {
            try
            {
                commander.Remove(owned[index]);
            }
            catch (Exception)
            {
                // Shutdown path: a route the table no longer has is the common reason to fail.
            }
        }
        _installed.Clear();
        SaveQuietly();
        return owned;
    }

    /// <summary>
    /// Reports whether a prefix is one of ours, which is what keeps a conflict check from treating
    /// this feature's own route as somebody else's.
    /// </summary>
    public bool Owns(string cidr) => _installed.Exists(route => route.Cidr == cidr);

    private void Forget(PeerEgressRoute route) => _installed.RemoveAll(existing => existing.Cidr == route.Cidr);

    private void SaveQuietly()
    {
        try
        {
            Save();
        }
        catch (Exception)
        {
            // The caller is already on a failure path and has an error to report.
        }
    }
}
